sn_node/src/put_validation.rs

```diff
@@ -758,33 +741,109 @@ impl Node {
-        } else {
-            debug!("Got a single valid spend for {unique_pubkey:?}");
-            Ok(verified_spends)
+        // return 2 spends max
+        let all_verified_spends: Vec<_> = all_verified_spends.into_iter().collect();
+        match all_verified_spends.as_slice() {
+            [one_spend] => Ok((one_spend.clone(), None)),
+            [one, two] => Ok((one.clone(), Some(two.clone()))),
+            [] => {
+                warn!("Invalid request: none of the spends were valid for {unique_pubkey:?}");
+                Err(Error::InvalidRequest(format!(
+                    "Found no valid spends while validating Spends for {unique_pubkey:?}"
+                )))
+            }
+            more => {
+                warn!("Got more than 2 verified spends, this might be a double spend spam attack, making sure to favour live branches (branches with spent descendants)");
+                let (one, two) = self.verified_spends_select_2_live(more).await?;
+                Ok((one, Some(two)))
+            }
+        }
+    }
+
+    async fn verified_spends_select_2_live(
+        &self,
+        many_spends: &[SignedSpend],
+    ) -> Result<(SignedSpend, SignedSpend)> {
+        // get all spends descendants
+        let mut tasks = JoinSet::new();
+        for spend in many_spends {
+            let descendants: BTreeSet<_> = spend
+                .spend
+                .spent_tx
+                .outputs
+                .iter()
+                .map(|o| o.unique_pubkey())
+                .map(SpendAddress::from_unique_pubkey)
+                .collect();
+            for d in descendants {
+                let self_clone = self.clone();
+                let spend_clone = spend.to_owned();
+                let _ = tasks.spawn(async move {
+                    let res = self_clone.network().get_raw_spends(d).await;
+                    (spend_clone, res)
+                });
+            }
+        }
+
+        // identify up to two live spends (aka spends with spent descendants)
+        let mut live_spends = BTreeSet::new();
+        while let Some(res) = tasks.join_next().await {
+            match res {
+                Ok((spend, Ok(_descendant))) => {
+                    let _inserted = live_spends.insert(spend);
+                    if live_spends.len() > 1 {
+                        let array: Vec<_> = live_spends.clone().into_iter().collect();
+                        if let [one, two] = array.as_slice() {
+                            warn!("Got two live spends {one:?} and {two:?}, things are messed up!");
+                            return Ok((one.to_owned().clone(), two.to_owned().clone()));
+                        }
+                    }
+                }
+                Ok((spend, Err(NetworkError::GetRecordError(GetRecordError::RecordNotFound)))) => {
+                    trace!("Spend {spend:?} descendant was not found, continuing...");
+                }
+                Ok((spend, Err(e))) => {
```

Comment:
Should we handle `GerRecordError::NotEnoughRecords`? An output could've just been spent and maybe did not have enough time to get to the majority nodes?

Reply:
Mmmh true, but if the parent is double spent, it should never become enough records as other nodes will reject it. I'd say we can safely ignore this one, what do you think?

Reply:
I was thinking about this situation where we have:
first spend: A -> B -> C utxo
double spend: A -> X utxo

Consider a bad actor spends A and sends it to B, and B being a good wallet sends it to itself (C). But before C got to enough nodes, the bad actor double spends A. 

Now, while checking, we will have 2 spends for A (double spend), but the live branch should be the first one. But if we don't consider the NotEnoughCopies err of C, then we might potentially lose it?

We might lose the live chain if: the bad actor spams double spend of A, so every node will store the first two `not_live_spends` which might not contain the A-B-C chain.

Reply:
Update: As discussed in slack, we'll be blindly trusting B to be valid even if 1 copy exists. That will open up more attacks. Resolving this.